lib/Sema/CSApply.cpp

```diff
@@ -7540,3 +7532,50 @@ void Solution::setExprTypes(Expr *expr) const {
   SetExprTypes SET(*this);
   expr->walk(SET);
 }
+
+/// MARK: SolutionResult implementation.
+
+SolutionResult SolutionResult::forSolved(Solution &&solution) {
+  SolutionResult result(Kind::Success);
+  result.solutions = new Solution(std::move(solution));
+  result.numSolutions = 1;
+  return result;
+}
+
+SolutionResult SolutionResult::forAmbiguous(
+    MutableArrayRef<Solution> solutions) {
+  assert(solutions.size() > 1 && "Not actually ambiguous");
+  SolutionResult result(Kind::Ambiguous);
+  result.solutions =
+      (Solution *)malloc(sizeof(Solution) * solutions.size());
```

Comment:
I think the life time of this array ends with constraint system which produced it, is that right? Maybe it makes sense to use constraint system allocator here...

Reply:
Yes, it does, but we're going to basically create one of these and move it around, so it doesn't seem worthwhile.

Reply:
Just so I better understand - why is malloc better in this case than just having it stack allocated or use constraint system allocator? At least we don't have to care about lifetime of the memory used for the list in either of these scenarios especially since it's just a container.

Reply:
I could just use an `std::vector` and avoid the whole issue; this should never be big

Reply:
That's why I'm curious what's the advantage of direct use of malloc in this case :)

Reply:
Pointless micro-optimization. Will fix